### ISSUE TITLE
Should `StateT` and `WriterT` have `Concurrent` instances?

### ISSUE THREAD

**Issue by changlinli** [typelevel/cats-effect#272]
Any monad transformer with writable state that doesn't have some sort of atomicity guarantee (such as a `ReaderT` of a `Ref`) has some weird `Concurrent` instances.

Take `StateT` for example.

```scala
import scala.concurrent.ExecutionContext.Implicits.global

type StateList[A] = StateT[IO, List[Int], A]

val stateOne = StateT[IO, List[Int], Int](x => IO((1 :: x, 1)))

val stateTwo = StateT[IO, List[Int], Int](x => IO((2 :: x, 2)))

val result = for {
  fiberX <- implicitly[Concurrent[StateList]].start(stateOne)
  fiberY <- implicitly[Concurrent[StateList]].start(stateTwo)
  x <- fiberX.join
  y <- fiberY.join
} yield (x, y)

// This is (List(2), (1, 2)) whereas I would've expected either (List(1, 2), (1, 2)) or (List(2, 1), (1, 2))
result.runEmpty.unsafeRunSync()
```

These instances are breaking what I would consider an intuitive expectation I have of these `Concurrent` instances which is that using `start` and `join` maintains serializability of state modifications, instead of dropping state modifications.

I think that this represents a potential head-scratcher for new users and would personally be in favor of removing these instances altogether. Thoughts?

**Comment by LukaJCB** [comment]
This is interesting and I agree with your premise. 

Also, for reference here's one Haskell package that also omits these instances for `MonadFork` which is very similar to our `Concurrent` http://hackage.haskell.org/package/monad-parallel-0.7.2.3/docs/Control-Monad-Parallel.html

**Comment by alexandru** [comment]
```scala
fa.start.flatMap(_.join) <-> fa
```

This is a law of `Concurrent`. If you want the StateT and WriterT instances to be dropped, then you need to come up with a law that excludes just them and that doesn’t contradict the above law, which I don’t think it is possible.

Also dropping the instances would be a breaking change between RC2 and 1.0.0 and we need a really good reason for breaking changes, other than feeling unintuitive or that Haskell doesn’t do it for MonadFork.

**Comment by djspiewak** [comment]
@changlinli This isn't really `Concurrent`'s fault so much as `StateT`'s. While intuitively it does feel like something has been "lost" when the second `join` happens, in reality the problem is that `State` just allows you to completely whack the threaded value. So when you do the second `join`, you are literally saying "ignore the state computed previous to this line and overwrite with the line of effects from the other fiber".

Conceptually, there's not much else that could be done, given the way that `State` is defined: you can *get the state* as a concrete value, and then use that value to run other effects. You asked those effects to be run in parallel (when you called `start`), which means the only state value it could possibly get is the one which was sequenced at the point where `start` happened. There's no way for that state to be reconciled with the other fiber, so you lose the results from the first `join`.

You *could* define a different sort of `State` which doesn't have this problem. For example:

```scala
final case class MonotoneStateT[F[_], S, A](run: F[(List[S => F[S]], A)]) {
  def flatMap[B](f: A => MonotoneStateT[F, S, B])(implicit F: Monad[F]): MonotoneStateT[F, S, B] = {
    val back = run flatMap {
      case (actions1, a) => 
        f(a).run flatMap {
          case (actions2, b) => (actions1 ++ actions2, b)
        }
    }

    MonotoneStateT(back)
  }

  def eval(init: S)(implicit F: Monad[F]): F[(S, A)] = {
    run flatMap {
      case (actions, a) => actions.foldRightM(init)(_(_)).strengthR(a)
    }
  }
}

object MonotoneStateT {

  def pure[F[_]: Applicative, S, A](a: A): MonotoneStateT[F, S, A] =
    MonotoneStateT(Applicative[F].pure((Nil, a)))

  def modifyF[F[_]: Applicative, S](f: S => F[S]): MonotoneStateT[F, S, Unit] =
    MonotoneStateT(Applicative[F].pure((f :: Nil, ())))

  def modify[F[_]: Applicative, S](f: S => S): MonotoneStateT[F, S, Unit] =
    modifyF(f.andThen(Applicative[F].pure(_)))
}
```

This would resolve the problem and you wouldn't have any "state loss", but it introduces a new issue: you can no longer define `get`. Since you don't have the ability to compute an `F[A]` from a specific value of `S`, there is no longer any interference between the state threading an the control flow. You can still run effects based on state, but those effects are *deferred* and will not be run in parallel when forked by `Concurrent` (since the `actions` are sequenced with `foldRightM` only when `eval` is run).

But this probably isn't what you want. 🙂

**Comment by changlinli** [comment]
@djspiewak I agree, but that's why I think `StateT` shouldn't have a `Concurrent` instance. There isn't anything wrong with `Concurrent`, `StateT` is just fundamentally sequential. Just as there isn't anything wrong with `Monad`, `Const` just doesn't have a valid `Monad` instance.

@alexandru The wider law I'm trying to express is hard to write down without some notion of decomposing IO actions into atomic units. I'm basically trying to say that for any effect that does writes, those writes (and corresponding reads) should be serializable, which I view as the usual criterion of lawful concurrency.

Here's a specialized version of that law. A no-op effect run concurrently should still be a no-op effect.

```
a.pure.start.flatMap(fiber => fa <* fiber.join) <-> fa

// Or equivalently in the way that I think syntactically it'll bite users
(for {
  fiber <- a.pure.start
  a <- fa
  _ <- fiber.join
} yield a) <-> fa
```

Another way of putting it is that it violates laws/invariants on the monad transformers themselves. An implicit law that I think a lot of people rely on and I've seen written but hasn't been codified is that `StateT` is equivalent to `ReaderT` if `put` is never used.

**Comment by changlinli** [comment]
Also on reflection `WriterT` might be fine since `join`s are sequential even if the effects are run concurrently.

**Comment by djspiewak** [comment]
Yes, `WriterT` should be fine. My `MonotoneStateT` is actually just `WriterT[F, List[S => F[S]], A]`.

I suppose in a sense having a `Concurrent` for `StateT` is analogous to having an `Effect` for `Kleisli` given a `Monoid[R]`. It's perfectly lawful, but it's almost certainly not going to be what you want under any circumstance. So I guess that's the argument, @alexandru: it's just not useful, and it allows newcomers to accidentally write programs which are going to behave very strangely. Obviously, with the run up to 1.0.0, breaking compatibility is a really tough sell, but I could see this potentially being a viable 1.1.0 change.

**Comment by changlinli** [comment]
Is deprecating it right away in `1.0.0` a viable choice?

@djspiewak There's actually a similar thing for the `Async` instance of `StateT` to your comment about the `Effect` for `Kleisli` with a `Monoid`. It injects a empty state with its `Monoid` instance which is also almost certainly not what you want.

**Comment by djspiewak** [comment]
@alexandru @changlinli Thinking about it, I would posit the following law:

```scala
(for {
  f <- fa.start
  _ <- fa
  a <- f.join
} yield a) <-> fa *> fa
```

I'm pretty sure this captures the intuition @changlinli is referencing. It will also be violated by `StateT`, but not violated by some other very useful types such as (obviously) `IO`.

As an aside, I wonder if we can generate side-effecting `IO` instances in the scalacheck generators. I was thinking about it, and types like `IO` are going to sort of vacuously pass this law due to the fact that the generators never actually sequence any effects (all of the generated `IO`s are, ultimately, pure). I think we can resolve this by putting a private `var` on `TestContext` of type (e.g.) `List[Int]`, then having a case in the `IO` generator which does something like this:

```scala
def genImpureApply[A: Arbitrary](implicit ec: TestContext): Gen[IO[A]] =
  arbitrary[A].map(a => IO(ec.effects ::= a))
```

Then, in the `Eq[IO[A]]`, we would split the current `ec.tick()` into two ticks: the first one runs the left `IO`, captures the value of `effects` and then sets it back to `Nil`, then runs the second tick which does the same. Comparing the value of `effects` would be part of the equality check.

That would allow tests like the above to pass *because `IO` satisfies the requirements*, not just because we aren't actually hitting these cases.

> If a lawful instance can be provided and if there are some use-cases for it, then we'd be inviting users to build incoherent instances and that's worse than us providing an instance that's seemingly not useful.

I agree with this point, which is part of why I think `Async[StateT[F, S, ?]]` for any `Async[F]` should still be supplied (it is both lawful and useful). Anyone defining their own `Concurrent` would end up causing ambiguities with that `Async`, so that's a pretty big red flag right there which would probably shoot down their attempt. If they choose to ignore it, then there's not really a lot we can do.

**Comment by changlinli** [comment]
@djspiewak Unfortunately I think `IO` violates your proposed law. Imagine a mutable counter that is non-atomically incremented (i.e. separate `get` and `set` commands rather than a single atomic `modify`) as your `fa`.

`fa *> fa` will always increase the counter by 2, whereas the concurrent version might only increase it by `1` depending on timing.

This is what I was referring to about need to classify operations as atomic if you want to have a more general law. Or more accurately you have to be able to identify what component actions are being considered for serializability. `pure` is trivially atomic so it works. 

With parametric polymorphism in the effect parameter of a monad transformer and the other laws, the `pure`-specialized version might be sufficiently constraining to capture the general intuition for all monad transformers, but I haven't thought enough about that to be sure.

**Comment by changlinli** [comment]
Also yeah I take back my `Async` comment about `StateT`. I was thinking about fs2's (ab)use of `LiftIO` and `Effect` to approximate `Concurrent`

**Comment by djspiewak** [comment]
@changlinli Hmm. We can improve the situation slightly with the following:

```scala
(for {
  f <- fa.start
  a1 <- fa
  a2 <- f.join
} yield a1 |+| a2) <-> (fa, fa).map2(_ |@| _)
```

It's not perfect – we can't say anything about the effects composing in the same way that the values have to – but it's something.

Ultimately, we're going to have this problem with `Concurrent` just in general. If people use non-monoidial effects in a concurrent setting, the results will be nondeterministic. That's just the nature of things.

**Comment by changlinli** [comment]
@djspiewak You can get around some of those problems by imposing conditions and one-way implications rather than trying to have unrestricted bidirectional equivalence. E.g. you might say if `fa` and `fb` are both terminating and deterministic (e.g. all valid evaluations of the action are always observationally equal) and given 

```scala
def concurrently(fa: F[A], fb: F[B]) = for {
  fiber <- fa.start
  b <- fb
  a <- fiber.join
} yield (a, b)

def sequentially(fa: F[B], fb: F[B]) = for {
  a <- fa
  b <-  fb
} yield (a, b)
```

if `concurrently(fa, fb) <-> concurrently (fb, fa)` then `sequentially(fa, fb) <-> seqentially(fb, fa)` (but not necessarily vice versa).

**Comment by djspiewak** [comment]
@changlinli Wouldn't it actually be `sequentially(fa, fb) <-> seqentially(fb, fa)` implies `concurrently(fa, fb) <-> concurrently (fb, fa)`, but not necessarily vice versa?

> In general you probably just want to assume determinism everywhere for all the effect laws, because otherwise even reflexive equality is hard to define.

I agree. The rabbit hole of nondeterminism is very very deep indeed. So what if we go back to the original variant of my proposed law and add the additional "soft" constraint, "All generated instances of your effect type `F[A]` must have independent effects (i.e. cannot close over some mutable cell shared with other generated effects)." If we have that soft constraint, then I think this law passes again:

```scala
(for {
  f <- fa.start
  _ <- fa
  a <- f.join
} yield a) <-> fa *> fa
```

WDYT?

**Comment by changlinli** [comment]
I actually do think it's `concurrently` implies `sequentially` because concurrent actions always subsume sequential actions. There is always a way of evaluating a concurrent pair of actions such that one occurs after another so any property of a deterministic set of concurrent actions must carry over into its sequential counterpart otherwise one possible evaluation of the concurrent action would differ from other possible evaluations.

WRT your law I think that works. I'd probably replace the "soft" constraint with just the constraint that the `for`-comprehension expression must be deterministic. It's a stronger condition, but it feels less hand-wavey to me. Granted getting down into the weeds of what it means for an effect to be deterministic is also liable to begin the hand-waving process, but it can be very non-obvious what would constitute a mutable cell (e.g. if you send read and write requests to a web API does that count as a mutable cell?).

**Comment by djspiewak** [comment]
@changlinli No, I'm pretty sure it's `sequential` implies `concurrent`. The reason for this is the implication is meant to catch non-commuative effects. Basically the law as I reformulated it asserts that "If `fa` and `fb` commute sequentially, then they should commute concurrently." We can't deterministically assert in the opposite direction since the concurrent commutativity check may have succeeded only by coincidence.

Either way, I think the weaker law with the soft restriction is a better way to go. I'm not sure what you mean by "the `for`-comprehension must be deterministic" though. Deterministic in what sense? Either way, we're still somewhat dependent on the `Arbitrary[F[A]]`, and so requiring (unenforcibly) that all such values are independent (which is the natural implementation regardless) will allow the law to deterministically pass.

**Comment by alexandru** [comment]
Guys, so what law do we need to add? Do you have consensus?

Can you make a PR?

**Comment by djspiewak** [comment]
I'm still in favor of what I proposed [in this comment](https://github.com/typelevel/cats-effect/issues/272#issuecomment-401859635). So absent further discussion, that's what I'd like to go with, but I don't want to just "win" by fiat. I definitely recognize the validity of @changlinli's points where he says we're still dependent on the `Arbitrary`, but I don't see a rigorous way around that in this case.

**Comment by changlinli** [comment]
My apologies for the late response, I haven't been as diligent about checking GitHub lately.

@alexandru To expand on the idea of concurrent implying sequential, what I meant is that, assuming you want deterministic invariants, if those invariants hold when executed concurrently, they must also hold when executed sequentially. That's because one possible runtime evaluation of a concurrent program is sequential evaluation, whereas, generally speaking, invariants of a sequential program go out the window when you evaluate its parts in parallel.

That is if you write 

```
val myAction = for {
  _ <- ioAction0.start
  _ <- ioAction1.start
} yield ()
```

if you want your code to have certain invariants under all legitimate evaluations of `myAction`, those same invariants must hold for 

```
val myAction1 = for {
  _ <- ioAction0
  _ <- ioAction1
} yield ()
```

because one possible runtime evaluation of `myAction` is equivalent to `myAction1` (e.g. if you have just one core and the scheduler just happens to put `ioAction1.start` after `ioAction.start`).

The opposite direction does not hold. There can definitely be invariants that `myAction1` respects that `myAction` does not.

That's why I said in a previous comment that `concurrently(fa, fb) <-> concurrently (fb, fa)` implies `sequentially(fa, fb) <-> seqentially(fb, fa)` but not vice versa.

@djspiewak I'm reluctant to go with "the cannot close over a mutable cell" restriction because it's not obvious what that might mean when talking about general effects (e.g. reading and writing to other a distributed set of services over the network). I definitely agree with what I think is the intuition you're getting at, namely that if you have independent, commutative effects then executing them concurrently should be the same as executing them sequentially. 

I think though that `concurrently(fa, fb) <-> concurrently (fb, fa)` implies `sequentially(fa, fb) <-> seqentially(fb, fa)` is a generalization of the independent effects condition. That is if effects are independent, then the implication holds. On the other hand even if `fa` and `fb` are not independent, I think the implication still holds for intuitively "correct" concurrent effects.

That being said it's still a style call. What exactly `<->` in `concurrently(fa, fb) <-> concurrently (fb, fa)` means is certainly a lot harder to pin down than the code might make it appear.

**Comment by alexandru** [comment]
@changlinli would you agree with us adding @djspiewak's proposed law in [his comment](https://github.com/typelevel/cats-effect/issues/272#issuecomment-401859635).

And who wants to work on a PR?

**Comment by changlinli** [comment]
@alexandru @djspiewak @LukaJCB I'm still not a big fan of that law, especially if you want that law to be re-used by other people writing their own `Concurrent` instances. It seems kind of onerous to enforce that people must structure their `Arbitrary` instances in a certain way otherwise that law will fail, all the more because this restricts users to provide the "easiest" `Arbitrary` instances possibles (e.g. instances which essentially don't do anything interesting concurrency-wise).

My proposed `concurrent(a, b) <-> concurrent(b, a)` implies `sequential(a, b) <-> sequential(b, a)` is also onerous because I can't think of a good way of doing it without either again forcing users to contort their `Arbitrary` instances or to use `==>` (at the cost of a large performance penalty).

Hence I would err on the side of having a more specialized, less elegant law, that removes the burden of having to craft "uninteresting" `Arbitrary` instances and allows users to use truly arbitrary `Arbitrary`s.

```
(for {
  fiber <- a.pure.start
  x <- fa
  _ <- fiber.join
} yield x) <-> fa
```

Roughly "doing nothing (`pure`) concurrently with another action is the same as just doing that action itself."

Thoughts?
